pybamm/citations.py

```diff
@@ -245,8 +247,28 @@ def print(self, filename=None, output_format="text", verbose=False):
                 f.write(citations)
 
 
+def register_citation(key):
+    """
+    Register a paper to be cited, one at a time.
+    It should be called only when the referenced functionality is actually being used.
+
+    .. warning::
+        Registering a BibTeX citation, with the same key as an existing citation,
+        will overwrite the current citation.
+
+    Parameters
+    ----------
+    key : str
+        - The citation key for an entry in `pybamm/CITATIONS.bib` or
+        - A BibTeX formatted citation
+    """
+    citations = Citations.citations
+    citations._register(key)
+
```

Comment:
I'm afraid that this leans towards being a breaking change for users, so we'll probably have to revert this or restructure the function to match the current API. We can't issue a breaking change without a warning message and a deprecation cycle across two/three releases being involved. Is it possible to stick with `pybamm.citations.register("<...>")` for now?

Reply:
I see. Yes, `pybamm.Citations.citations.register(key)` is also a bit messy, and the try-except block was also masking the import error, which we probably don't want (because it was masking any other `Exception`s too). I think a better way would be to rethink the citations workflow from start to end. The current one is using `self.read_citations()` during `__init__`, and we are importing the class itself when we run `import pybamm` – here we could make the `citations.register` functionality optional across the codebase, which can work, but then we want to encourage people to cite their references properly and not make it optional to do so. One way out is to make `pybtex` a mandatory dependency and remove the `[cite]` extra – which should be the last way out.

One option that seems reasonable at the time of writing is to put import guards across all of the `pybtex`-related functionality. Wherever we are using `import_optional_dependency("pybtex")`, if `pybtex` is not present, we are currently not implementing any try-except blocks to ensure that the `pybtex`-specific functions are not called. Maybe doing so and adding a helpful error message can fix it? i.e., `read_citations`, `_add_citation`, and `_parse_citation`, and `print` should not run if `pybtex` is not installed.

Reply:
All right, yes we can definitely guard `pybtex` imports for the moment :)
I changed this in e02662f